gunrock/app/lp/lp_enactor.cuh

```diff
@@ -0,0 +1,647 @@
+    this->max_num_value__associates = 0;
+  }
+
+  /**
+   * @brief LPEnactor destructor
+   */
+  virtual ~Enactor() {
+    // Release();
+  }
+
+  /*
+   * @brief Releasing allocated memory space
+   * @param target The location to release memory from
+   * \return cudaError_t error message(s), if any
+   */
+  cudaError_t Release(util::Location target = util::LOCATION_ALL) {
+    cudaError_t retval = cudaSuccess;
+    GUARD_CU(BaseEnactor::Release(target));
+    delete[] iterations;
+    iterations = NULL;
+    problem = NULL;
+    return retval;
+  }
+
+  /**
+   * @brief Initialize the enactor.
+   * @param[in] problem The problem object.
+   * @param[in] target Target location of data
+   * \return cudaError_t error message(s), if any
+   */
+  cudaError_t Init(Problem &problem, util::Location target = util::DEVICE) {
+    cudaError_t retval = cudaSuccess;
+    this->problem = &problem;
+    util::Parameters &parameters = problem.parameters;
+
+    GUARD_CU(BaseEnactor::Init(problem, Enactor_None, 2, NULL, target, false));
+
+    for (int gpu = 0; gpu < this->num_gpus; gpu++) {
+      
+      GUARD_CU(util::SetDevice(this->gpu_idx[gpu]));
+
+      // for each gpu the slice is initialised to the starting point (peer = 0)
+      auto &enactor_slice = this->enactor_slices[gpu * this->num_gpus + 0];
+      auto &graph = problem.sub_graphs[gpu];
+      
+      // TODO
+      // does this allocate function duplicate data on each gpu?
+      // no it is agnostic to that as the partition should have handled that 
+      // and anyway we call this once for each gpu not for each enactor slice
+      GUARD_CU(enactor_slice.frontier.Allocate(graph.nodes, graph.edges,
+                                               this->queue_factors));
+      
+      // this is where we will have the enactor slice data allocated
+      // not allocated but pointer referenced
+
+      // each enactor slice's operator parameter labels have data slices belonging to the specific gpu
+      for (int peer = 0; peer < this->num_gpus; peer++) {
+        this->enactor_slices[gpu * this->num_gpus + peer]
+            .oprtr_parameters.labels = &(problem.data_slices[gpu]->labels);
+      }
+    }
+
+    iterations = new IterationT[this->num_gpus];
+    for (int gpu = 0; gpu < this->num_gpus; gpu++) {
+      GUARD_CU(iterations[gpu].Init(this, gpu));
+    }
+
+    GUARD_CU(this->Init_Threads(
+        this, (CUT_THREADROUTINE) & (GunrockThread<EnactorT>)));
+    return retval;
+  }
+
+  /**
+   * @brief Reset enactor
+   * @param[in] src Source node to start primitive.
+   * @param[in] target Target location of data
+   * \return cudaError_t error message(s), if any
+   */
+  cudaError_t Reset(VertexT src, util::Location target = util::DEVICE) {
+    typedef typename GraphT::GpT GpT;
+    cudaError_t retval = cudaSuccess;
+    GUARD_CU(BaseEnactor::Reset(target));
+
+    SizeT num_nodes = this->problem->data_slices[0][0].sub_graph[0].nodes;
```

Comment:
Hard-coding [0][0] here worries me a bit. Have you tried this on a machine with multiple GPUs, changing the --device you use? Will these [0][0] accesses always be valid?